File: anagrams_2012-07-18/anagrams.rb
#!/usr/bin/env ruby

f = File.new('/usr/share/dict/words')
all_words = f.read()

all_anagrams = []
f.seek(0)
f.readlines.each_with_index do |word, i|
  anagrams = []
  letters = word.strip.downcase.chars.to_a
  if letters.length < 15
    permutations = letters.permutation.to_a.uniq
    puts "#{i} #{word.strip}: #{permutations.length} potential anagrams..."
    permutations.each do |p|
      w = p.join
      re = Regexp.new("^#{w}$", Regexp::IGNORECASE + Regexp::MULTILINE)
      anagrams << w if all_words.match(re)
    end
    all_anagrams << anagrams.sort if anagrams.length > 1
  end
  break if i > 10
end

all_anagrams = all_anagrams.uniq.sort { |a,b| b.length <=> a.length }
all_anagrams.each do |a|
  puts a.inspect
end

File: anagrams_2012-07-18/anagrams.py
#!/usr/bin/env python

import itertools
import re

f = open('/usr/share/dict/words')
all_words = f.read()

all_anagrams = []
f.seek(0)
for i, word in enumerate(f.readlines()):
  anagrams = []
  word = word.strip().lower()
  if len(word) < 5:
    permutations = set(itertools.permutations(word))
    num_perms = len(permutations)
    print '%(i)s %(word)s: %(num_perms)s potential anagrams...' % locals()
    for p in permutations:
      w = ''.join(p)
      if re.search('^%s$' % w, all_words, re.IGNORECASE + re.MULTILINE):
        anagrams.append(w)
    if len(anagrams) > 1:
      all_anagrams.append( tuple(sorted(anagrams)) ) 
  if i > 1000: break

all_anagrams = sorted(set(all_anagrams), lambda a,b: cmp(len(b), len(a)))
for a in all_anagrams:
  print a
